crates/bevy_ecs/src/system/builder.rs

```diff
@@ -337,6 +339,194 @@ unsafe impl<'s, T: FromWorld + Send + 'static> SystemParamBuilder<Local<'s, T>>
+    /// Creates a [`SystemParamBuilder`] for a [`FilteredResources`] that accepts a callback to configure the [`FilteredResourcesBuilder`].
+    /// This boxes the callback so that it has a common type.
+    pub fn new_box(f: impl FnOnce(&mut FilteredResourcesBuilder) + 'a) -> Self {
+        Self(Box::new(f))
+    }
+}
+
+// SAFETY: Calls `Access::build`.
+unsafe impl<'w, T: FnOnce(&mut FilteredResourcesBuilder)> SystemParamBuilder<FilteredResources<'w>>
+    for FilteredResourcesParamBuilder<T>
+{
+    fn build(
+        self,
+        world: &mut World,
+        system_meta: &mut SystemMeta,
+    ) -> <FilteredResources<'w> as SystemParam>::State {
+        let mut builder = FilteredResourcesBuilder::new(world);
+        (self.0)(&mut builder);
+        let access = builder.build();
+        SystemParamBuilder::<FilteredResources>::build(access, world, system_meta)
+    }
+}
+
+// SAFETY: Resource ComponentId and ArchetypeComponentId access is applied to SystemMeta. If this FilteredResources
+// conflicts with any prior access, a panic will occur.
+unsafe impl<'w> SystemParamBuilder<FilteredResources<'w>> for Access<ComponentId> {
+    fn build(
+        self,
+        world: &mut World,
+        meta: &mut SystemMeta,
+    ) -> <FilteredResources<'w> as SystemParam>::State {
+        // Accessing all resources can't be done safely, so we need to prohibit it.
+        // The problem is that we won't know the `ArchetypeComponentId` for resources added in the future.
+        // For components, that can be handled with `SystemParam::new_archetype()`, but there is no equivalent for resources.
+        if self.has_read_all_resources() {
+            let system_name = &meta.name;
+            panic!("FilteredResources in system {system_name} accesses all resources. This is not supported.");
+        }
+
+        let combined_access = meta.component_access_set.combined_access();
+        let conflicts = combined_access.get_conflicts(&self);
+        if !conflicts.is_empty() {
+            let accesses = match conflicts {
+                AccessConflicts::All => "",
+                AccessConflicts::Individual(indices) => &format!(
+                    " {}",
+                    indices
+                        .ones()
+                        .map(|index| world
+                            .components
+                            .get_info(ComponentId::get_sparse_set_index(index))
+                            .unwrap()
+                            .name())
+                        .collect::<Vec<&str>>()
+                        .join(", ")
+                ),
+            };
+            let system_name = &meta.name;
+            panic!("error[B0002]: FilteredResources in system {system_name} accesses resources(s){accesses} in a way that conflicts with a previous system parameter. Consider removing the duplicate access. See: https://bevyengine.org/learn/errors/#b0002");
+        }
```

Comment:
Could this code be extracted in a function? It almost duplicated in the impl for `FilteredResourcesMutBuilder` and in `Query`'s `assert_component_access_compatibility`.